codgen: Handle neg. exp. in create_expand_pow_optimization

bjodah opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes gh-15335

#### Brief description of what is fixed or changed
~~We probably still need a subclass of Mul which doesn't "auto-evaluate" to fix remaining issues.~~ (no longer needed thanks to @smichr adding `UnevaluatedExpr`)

#### Other comments


#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* codegen
  * `create_expand_pow_optimization` is now more robust
<!-- END RELEASE NOTES -->

bjodah committed:
codgen: Handle neg. exp. in create_expand_pow_optimization

Partially fixes gh-15335

smichr commented:
> We probably still need a subclass of Mul which doesn't "auto-evaluate" to fix remaining issues.

```python
>>> from sympy.codegen.rewriting import *
>>> opt4 = create_expand_pow_optimization(4)
>>> ccode(optimize(-x**4, [opt4]))
'-x*x*x*x'
>>> ccode(optimize((x**4 - x**2), [opt4]))
'-x*x + x*x*x*x'
```

with

```diff
diff --git a/sympy/codegen/rewriting.py b/sympy/codegen/rewriting.py
index ef99f22..db02e81 100644
--- a/sympy/codegen/rewriting.py
+++ b/sympy/codegen/rewriting.py
@@ -35,6 +35,7 @@
 from sympy import log, exp, Max, Min, Wild, expand_log, Dummy
 from sympy.codegen.cfunctions import log1p, log2, exp2, expm1
 from sympy.core.mul import Mul
+from sympy.core.expr import UnevaluatedExpr
 from sympy.utilities.iterables import sift


@@ -221,7 +222,7 @@ def create_expand_pow_optimization(limit):
     """
     return ReplaceOptim(
         lambda e: e.is_Pow and e.base.is_symbol and e.exp.is_integer and e.exp.is_nonnegative and e.exp <= limit,
-        lambda p: Mul(*([p.base]*p.exp), evaluate=False)
+        lambda p: UnevaluatedExpr(Mul(*([p.base]*p.exp), evaluate=False))
     )

 # Collections of optimizations:```

smichr committed:
use UnevaluatedExpr in ccode optimization

It is also important to use Integer rather than integer
since var('x', integer=True).is_integer is True...but
the routine is not meant to handle symbolic integers.

bjodah committed:
Merge pull request #16 from smichr/cc

use UnevaluatedExpr in ccode optimization

bjodah committed:
create_expand_pow_optimization: fix docstring & another test

bjodah commented:
Thanks a lot! I've added your edits and updated the PR description. Hopefully this fixes the original issue completely.
